File: deps/v8/test/mjsunit/shared-memory/mutex.js
// Copyright 2022 the V8 project authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
//
// Flags: --harmony-struct

let mutex = new Atomics.Mutex;
let locked_count = 0;

assertEquals(42, Atomics.Mutex.lock(mutex, () => {
  locked_count++; return 42;
}));
assertEquals(locked_count, 1);

let lockResult = Atomics.Mutex.tryLock(mutex, () => {
  locked_count++;
});
assertEquals(true, lockResult.success);
assertEquals(locked_count, 2);

// Recursively locking throws.
Atomics.Mutex.lock(mutex, () => {
  locked_count++;
  assertThrows(() => {
    Atomics.Mutex.lock(mutex, () => { throw "unreachable"; });
  }, Error);
});
assertEquals(locked_count, 3);

// Recursive tryLocking returns returns false success value.
Atomics.Mutex.lock(mutex, () => {
  locked_count++;
  lockResult = Atomics.Mutex.tryLock(mutex, () => {
    throw 'unreachable';
  });
  assertEquals(false, lockResult.success);
  assertEquals(undefined, lockResult.value);
});
assertEquals(locked_count, 4);

// Throwing in the callback should unlock the mutex.
assertThrowsEquals(() => {
  Atomics.Mutex.lock(mutex, () => { throw 42; });
}, 42);
Atomics.Mutex.lock(mutex, () => { locked_count++; });
assertEquals(locked_count, 5);

assertThrowsEquals(() => {
  Atomics.Mutex.tryLock(mutex, () => { throw 42; });
}, 42);
Atomics.Mutex.tryLock(mutex, () => { locked_count++; });
assertEquals(locked_count, 6);

// Mutexes can be assigned to shared objects.
(function TestMutexCanBeAssignedToSharedObjects() {
  const Box = new SharedStructType(["payload"]);
  const box = new Box;
  box.payload = mutex;
})();

// lockWithTimeout's return value is an object with a value field
// and a success field.
lockResult = Atomics.Mutex.lockWithTimeout(mutex, () => 42, 1);
assertEquals(42, lockResult.value);
assertEquals(true, lockResult.success);

// Timeout must be a number.
assertThrows(() => {
  Atomics.Mutex.lockWithTimeout(mutex, () => {}, '42');
})
assertThrows(() => {
  Atomics.Mutex.lockWithTimeout(mutex, () => {}, {});
})
assertThrows(() => {
  Atomics.Mutex.lockWithTimeout(mutex, () => {});
})

File: deps/v8/src/base/platform/mutex.cc
// Copyright 2013 the V8 project authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "src/base/platform/mutex.h"

#include "src/base/platform/platform.h"

namespace v8 {
namespace base {

RecursiveMutex::~RecursiveMutex() {
  DCHECK_EQ(0, level_);
}

void RecursiveMutex::Lock() {
  int own_id = v8::base::OS::GetCurrentThreadId();
  if (thread_id_ == own_id) {
    level_++;
    return;
  }
  mutex_.Lock();
  DCHECK_EQ(0, level_);
  thread_id_ = own_id;
  level_ = 1;
}

void RecursiveMutex::Unlock() {
#ifdef DEBUG
  int own_id = v8::base::OS::GetCurrentThreadId();
  CHECK_EQ(thread_id_, own_id);
#endif
  if ((--level_) == 0) {
    thread_id_ = 0;
    mutex_.Unlock();
  }
}

bool RecursiveMutex::TryLock() {
  int own_id = v8::base::OS::GetCurrentThreadId();
  if (thread_id_ == own_id) {
    level_++;
    return true;
  }
  if (mutex_.TryLock()) {
    DCHECK_EQ(0, level_);
    thread_id_ = own_id;
    level_ = 1;
    return true;
  }
  return false;
}

Mutex::Mutex() {
#ifdef DEBUG
  level_ = 0;
#endif
}

Mutex::~Mutex() { DCHECK_EQ(0, level_); }

void Mutex::Lock() ABSL_NO_THREAD_SAFETY_ANALYSIS {
  native_handle_.Lock();
  AssertUnheldAndMark();
}

void Mutex::Unlock() ABSL_NO_THREAD_SAFETY_ANALYSIS {
  AssertHeldAndUnmark();
  native_handle_.Unlock();
}

bool Mutex::TryLock() ABSL_NO_THREAD_SAFETY_ANALYSIS {
  if (!native_handle_.TryLock()) return false;
  AssertUnheldAndMark();
  return true;
}

}  // namespace base
}  // namespace v8
